File: app/src/main/java/com/example/attendancesystem/models/QRCodeData.kt
package com.example.attendancesystem.models

import com.google.gson.Gson
import com.google.gson.JsonSyntaxException
import java.util.*

data class QRCodeData(
    val teacherId: String,
    val sessionId: String,
    val userId: String,
    val timestamp: Long,
    val scheduleId: String,
    val subject: String,
    val section: String,
    val expirationMinutes: Int = DEFAULT_EXPIRATION_MINUTES,
    val latitude: Double? = null,
    val longitude: Double? = null,
    val locationTimestamp: Long? = null
) {
    companion object {
        const val DEFAULT_EXPIRATION_MINUTES = 5

        fun createWithExpiration(
            teacherId: String,
            sessionId: String,
            userId: String,
            scheduleId: String,
            subject: String,
            section: String,
            expirationMinutes: Int = DEFAULT_EXPIRATION_MINUTES,
            latitude: Double? = null,
            longitude: Double? = null
        ): QRCodeData {
            return QRCodeData(
                teacherId = teacherId,
                sessionId = sessionId,
                userId = userId,
                timestamp = System.currentTimeMillis(),
                scheduleId = scheduleId,
                subject = subject,
                section = section,
                expirationMinutes = expirationMinutes,
                latitude = latitude,
                longitude = longitude,
                locationTimestamp = if (latitude != null && longitude != null) System.currentTimeMillis() else null
            )
        }

        fun fromJson(json: String): QRCodeData {
            return try {
                Gson().fromJson(json, QRCodeData::class.java)
            } catch (e: JsonSyntaxException) {
                throw IllegalArgumentException("Invalid QR code data format")
            }
        }
    }

    fun toJson(): String {
        return Gson().toJson(this)
    }

    fun isExpired(): Boolean {
        val expirationTime = timestamp + (expirationMinutes * 60 * 1000)
        return System.currentTimeMillis() > expirationTime
    }

    fun getRemainingTimeInMillis(): Long {
        val expirationTime = timestamp + (expirationMinutes * 60 * 1000)
        val remaining = expirationTime - System.currentTimeMillis()
        return if (remaining > 0) remaining else 0
    }

    fun hasLocation(): Boolean {
        return latitude != null && longitude != null
    }

    fun getLocationData(): com.example.attendancesystem.utils.LocationManager.LocationData? {
        return if (hasLocation()) {
            com.example.attendancesystem.utils.LocationManager.LocationData(
                latitude = latitude!!,
                longitude = longitude!!,
                timestamp = locationTimestamp ?: timestamp
            )
        } else {
            null
        }
    }
}

File: flutter_app/lib/core/models/qr_code_data.dart
import 'dart:convert';

class QRCodeData {
  final String teacherId;
  final String sessionId;
  final String userId;
  final int timestamp;
  final String scheduleId;
  final String subject;
  final String section;
  final int expirationMinutes;
  final double? latitude;
  final double? longitude;
  final int? locationTimestamp;

  static const int defaultExpirationMinutes = 5;

  QRCodeData({
    required this.teacherId,
    required this.sessionId,
    required this.userId,
    required this.timestamp,
    required this.scheduleId,
    required this.subject,
    required this.section,
    this.expirationMinutes = defaultExpirationMinutes,
    this.latitude,
    this.longitude,
    this.locationTimestamp,
  });

  factory QRCodeData.createWithExpiration({
    required String teacherId,
    required String sessionId,
    required String userId,
    required String scheduleId,
    required String subject,
    required String section,
    int expirationMinutes = defaultExpirationMinutes,
    double? latitude,
    double? longitude,
  }) {
    return QRCodeData(
      teacherId: teacherId,
      sessionId: sessionId,
      userId: userId,
      timestamp: DateTime.now().millisecondsSinceEpoch,
      scheduleId: scheduleId,
      subject: subject,
      section: section,
      expirationMinutes: expirationMinutes,
      latitude: latitude,
      longitude: longitude,
      locationTimestamp: (latitude != null && longitude != null)
          ? DateTime.now().millisecondsSinceEpoch
          : null,
    );
  }

  factory QRCodeData.fromJson(String jsonString) {
    try {
      final Map<String, dynamic> data = json.decode(jsonString);
      return QRCodeData(
        teacherId: data['teacherId'] ?? '',
        sessionId: data['sessionId'] ?? '',
        userId: data['userId'] ?? '',
        timestamp: data['timestamp'] ?? 0,
        scheduleId: data['scheduleId'] ?? '',
        subject: data['subject'] ?? '',
        section: data['section'] ?? '',
        expirationMinutes: data['expirationMinutes'] ?? defaultExpirationMinutes,
        latitude: data['latitude']?.toDouble(),
        longitude: data['longitude']?.toDouble(),
        locationTimestamp: data['locationTimestamp'],
      );
    } catch (e) {
      throw ArgumentError('Invalid QR code data format');
    }
  }

  String toJson() {
    return json.encode({
      'teacherId': teacherId,
      'sessionId': sessionId,
      'userId': userId,
      'timestamp': timestamp,
      'scheduleId': scheduleId,
      'subject': subject,
      'section': section,
      'expirationMinutes': expirationMinutes,
      if (latitude != null) 'latitude': latitude,
      if (longitude != null) 'longitude': longitude,
      if (locationTimestamp != null) 'locationTimestamp': locationTimestamp,
    });
  }

  bool get isExpired {
    final expirationTime = timestamp + (expirationMinutes * 60 * 1000);
    return DateTime.now().millisecondsSinceEpoch > expirationTime;
  }

  int get remainingTimeInMillis {
    final expirationTime = timestamp + (expirationMinutes * 60 * 1000);
    final remaining = expirationTime - DateTime.now().millisecondsSinceEpoch;
    return remaining > 0 ? remaining : 0;
  }

  bool get hasLocation => latitude != null && longitude != null;
}
